Sources/kai/Checker.swift

```diff
@@ -1344,6 +1349,65 @@ extension Checker {
                 checkStmt(stmt)
             }
 
+        case .stmtSwitch(let subject, let cases, _):
+            let subjectType: Type?
+            
+            if let subject = subject {
+                subjectType = checkExpr(subject)
+            } else {
+                subjectType = nil
+            }
+            
+            var hasDefaultCase: Bool = false
+            
+            for caseStmt in cases {
+                guard case .stmtCase(let match, let body, _) = caseStmt else {
+                    reportError("Expected `case` block in `switch`, got `\(caseStmt)`", at: caseStmt)
+                    return
+                }
+                
+                guard !hasDefaultCase else {
+                    reportError("Additional `case` blocks cannot be after a `default` block", at: caseStmt)
+                    return
+                }
+                
+                if let match = match {
+                    let matchType = checkExpr(match)
+                    
+                    if let subjectType = subjectType {
+                        guard canImplicitlyConvert(matchType, to: subjectType) else {
+                            reportError("Cannot implicitly convert type `\(matchType)` to `\(subjectType)`", at: match)
+                            return
+                        }
+                    } else /* booleanesque */ {
+                        guard canImplicitlyConvert(matchType, to: Type.bool) else {
+                            reportError("Non-bool `\(match)` (type `\(matchType)`) used as condition", at: match)
+                            return
+                        }
+                    }
+                    
+                    let bodyScope = pushScope(for: body)
+                    defer { popScope() }
+                    
+                    guard case .stmtBlock(let stmts, _) = body else {
+                        panic()
+                    }
+                    
+                    // TODO(Brett, vdka): check for returns and handle all that
+                    // logic.
```

Comment:
We should actually change the way `return`'s are checked for type correctness, I can do that on a main branch and that will make this code sufficient.